src/boot_O2_g3/fault.c

```diff
@@ -572,354 +696,441 @@ void Fault_DrawMemDump(u32 pc, u32 sp, u32 unk0, u32 unk1) {
-        sFaultContext->faultedThread = faultedThread;
-        while (!sFaultContext->faultHandlerEnabled) {
+        __osSetFpcCsr(__osGetFpcCsr() & ~(FPCSR_EV | FPCSR_EZ | FPCSR_EO | FPCSR_EU | FPCSR_EI));
+        sFaultInstance->faultedThread = faultedThread;
+
+        while (!sFaultInstance->faultHandlerEnabled) {
             Fault_Sleep(1000);
         }
-        Fault_Sleep(500);
-        Fault_CommitFB();
+        Fault_Sleep(1000 / 2);
 
-        if (sFaultContext->faultActive) {
+        // Show fault framebuffer
+        Fault_DisplayFrameBuffer();
+
+        if (sFaultInstance->autoScroll) {
             Fault_Wait5Seconds();
         } else {
-            Fault_DrawCornerRec(0xF801);
+            // Draw error bar signifying the crash screen is available
+            Fault_DrawCornerRec(GPACK_RGBA5551(255, 0, 0, 1));
             Fault_WaitForButtonCombo();
         }
 
-        sFaultContext->faultActive = 1;
-        FaultDrawer_SetForeColor(0xFFFF);
-        FaultDrawer_SetBackColor(0);
+        // Set auto-scrolling and default colors
+        sFaultInstance->autoScroll = true;
+        FaultDrawer_SetForeColor(GPACK_RGBA5551(255, 255, 255, 1));
+        FaultDrawer_SetBackColor(GPACK_RGBA5551(0, 0, 0, 0));
 
+        // Draw pages
         do {
+            // Thread context page
             Fault_PrintThreadContext(faultedThread);
             osSyncPrintfThreadContext(faultedThread);
             Fault_WaitForInput();
+
+            // Stack trace page
             Fault_DrawStackTrace(faultedThread, 0);
-            osSyncPrintfStackTrace(faultedThread, 0);
+            Fault_LogStackTrace(faultedThread, 0);
             Fault_WaitForInput();
+
+            // Client pages
             Fault_ProcessClients();
+
+            // Memory dump page
             Fault_DrawMemDump((u32)(faultedThread->context.pc - 0x100), (u32)faultedThread->context.sp, 0, 0);
             Fault_DrawStackTrace(faultedThread, 1);
-            osSyncPrintfStackTrace(faultedThread, 1);
+            Fault_LogStackTrace(faultedThread, 1);
             Fault_WaitForInput();
+
+            // End page
             Fault_FillScreenRed();
-            FaultDrawer_DrawText(0x40, 0x50, "    CONGRATURATIONS!    ");
-            FaultDrawer_DrawText(0x40, 0x5A, "All Pages are displayed.");
-            FaultDrawer_DrawText(0x40, 0x64, "       THANK YOU!       ");
-            FaultDrawer_DrawText(0x40, 0x6E, " You are great debugger!");
+            FaultDrawer_DrawText(64, 80, "    CONGRATURATIONS!    ");
+            FaultDrawer_DrawText(64, 90, "All Pages are displayed.");
+            FaultDrawer_DrawText(64, 100, "       THANK YOU!       ");
+            FaultDrawer_DrawText(64, 110, " You are great debugger!");
             Fault_WaitForInput();
+        } while (!sFaultInstance->exit);
 
-        } while (!sFaultContext->exitDebugger);
-
-        while (!sFaultContext->exitDebugger) {
+        while (!sFaultInstance->exit) {
```

Comment:
```suggestion
        while (!sFaultInstance->exit) {}
```
As well as remove the `;` and `}` below.